File: src-backend/routes/appointments.js
const express = require('express');
const { supabase } = require('../db/client');
const { getMetaConfig, sendWhatsAppMessage } = require('../services/meta-service');
const { normalizePhone } = require('../services/phone-normalizer');

const router = express.Router();

// 1. Obtener todas las citas (con JOIN de contactos)
router.get('/', async (req, res) => {
  try {
    const { data: appointments, error } = await supabase
      .from('appointments')
      .select(`
        *,
        contacts (
          name,
          phone,
          var1,
          var2
        )
      `)
      .order('starts_at', { ascending: false });

    if (error) throw error;

    // Aplanamos la respuesta para mantener la compatibilidad con tu estructura anterior
    const formattedAppointments = appointments.map(app => ({
      ...app,
      contact_name: app.contacts ? app.contacts.name : null,
      contact_phone: app.contacts ? app.contacts.phone : null,
      var1: app.contacts ? app.contacts.var1 : null,
      var2: app.contacts ? app.contacts.var2 : null
    }));

    res.json({ appointments: formattedAppointments });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

// 2. Crear una nueva cita (e insertar contacto si no existe)
router.post('/', async (req, res) => {
  const { phone, name, service, starts_at, ends_at, defaultCountry } = req.body;
  const normalized = normalizePhone(phone, defaultCountry || '+57');
  
  if (!normalized.valid) {
    return res.status(400).json({ error: normalized.error });
  }
  
  const cleanPhone = normalized.phone;

  try {
    // Buscar si el contacto ya existe por su teléfono
    let { data: contact, error: contactError } = await supabase
      .from('contacts')
      .select('*')
      .eq('phone', cleanPhone)
      .maybeSingle(); // Retorna null si no encuentra filas en vez de lanzar error

    if (contactError) throw contactError;

    // Si no existe, lo creamos
    if (!contact) {
      const { data: newContact, error: insertContactError } = await supabase
        .from('contacts')
        .insert([{ phone: cleanPhone, name, source: 'appointment' }])
        .select()
        .single();

      if (insertContactError) throw insertContactError;
      contact = newContact;
    }

    // Insertar la cita asignando el contact_id correspondiente
    const { data: newAppointment, error: appError } = await supabase
      .from('appointments')
      .insert([
        {
          contact_id: contact.id,
          service,
          starts_at,
          ends_at,
          status: 'scheduled'
        }
      ])
      .select()
      .single();

    if (appError) throw appError;

    // Enviar notificación por WhatsApp
    const config = await getMetaConfig();
    await sendWhatsAppMessage({
      access_token: config.access_token,
      phone_number_id: config.phone_number_id,
      to: cleanPhone,
      templateName: 'utility_confirmation',
      language: 'es',
      components: [
        { 
          type: 'body', 
          parameters: [
            { type: 'text', text: name }, 
            { type: 'text', text: service }, 
            { type: 'text', text: starts_at }
          ] 
        }
      ]
    });

    res.json({ appointmentId: newAppointment.id });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

// 3. Confirmar una cita cambiándole el estado
router.post('/:id/confirm', async (req, res) => {
  const { id } = req.params;
  try {
    const { error } = await supabase
      .from('appointments')
      .update({ status: 'confirmed', updated_at: new Date().toISOString() })
      .eq('id', id);

    if (error) throw error;

    res.json({ ok: true });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

module.exports = router;

File: src-backend/routes/appointments.py
from fastapi import APIRouter, HTTPException, status
from pydantic import BaseModel
from typing import Optional
from db.client import db_client
from services.phone_normalizer import normalize_phone
from services.meta_service import get_meta_config, send_whatsapp_template_message

router = APIRouter(prefix="/api/appointments", tags=["appointments"])

class CreateAppointmentSchema(BaseModel):
    phone: str
    name: str
    service: str
    starts_at: str
    ends_at: str
    defaultCountry: Optional[str] = "+57"

@router.get("")
async def get_appointments_endpoint():
    """
    Lista las citas agendadas integrando los datos del contacto asociado.
    """
    try:
        query = """
            SELECT a.*, c.name AS contact_name, c.phone AS contact_phone, c.var1, c.var2
            FROM appointments a
            LEFT JOIN contacts c ON c.id = a.contact_id
            ORDER BY a.starts_at DESC
        """
        appointments = await db_client.all(query)
        return {"appointments": appointments}
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))

@router.post("")
async def create_appointment_endpoint(payload: CreateAppointmentSchema):
    """
    Crea una nueva cita, resolviendo o registrando el contacto y enviando el mensaje de confirmación.
    """
    # 1. Normalizar el teléfono
    norm = normalize_phone(payload.phone, payload.defaultCountry)
    if not norm["valid"]:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=norm["error"])
        
    clean_phone = norm["phone"]
    
    try:
        # 2. Buscar o crear el contacto
        contact = await db_client.get("SELECT * FROM contacts WHERE phone = ?", (clean_phone,))
        if not contact:
            insert_contact_res = await db_client.run(
                """
                INSERT INTO contacts (phone, name, source, created_at, updated_at)
                VALUES (?, ?, 'appointment', datetime('now'), datetime('now'))
                """,
                (clean_phone, payload.name)
            )
            contact_id = insert_contact_res["lastID"]
        else:
            contact_id = contact["id"]
            
        # 3. Insertar la cita
        insert_app_res = await db_client.run(
            """
            INSERT INTO appointments (contact_id, service, starts_at, ends_at, status, created_at, updated_at)
            VALUES (?, ?, ?, ?, 'scheduled', datetime('now'), datetime('now'))
            """,
            (contact_id, payload.service, payload.starts_at, payload.ends_at)
        )
        appointment_id = insert_app_res["lastID"]
        
        # 4. Enviar notificación de confirmación por WhatsApp en segundo plano
        config = await get_meta_config()
        if config.get("access_token") and config.get("phone_number_id"):
            components = [
                {
                    "type": "body",
                    "parameters": [
                        {"type": "text", "text": payload.name},
                        {"type": "text", "text": payload.service},
                        {"type": "text", "text": payload.starts_at}
                    ]
                }
            ]
            try:
                await send_whatsapp_template_message(
                    access_token=config["access_token"],
                    phone_number_id=config["phone_number_id"],
                    to=clean_phone,
                    template_name="utility_confirmation",
                    language="es",
                    components=components
                )
            except Exception as meta_err:
                print(f"Error enviando mensaje de WhatsApp: {meta_err}")
                
        return {"appointmentId": appointment_id}
        
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))

@router.post("/{appointment_id}/confirm")
async def confirm_appointment_endpoint(appointment_id: int):
    """
    Confirma una cita de manera manual.
    """
    try:
        await db_client.run(
            "UPDATE appointments SET status = 'confirmed', updated_at = datetime('now') WHERE id = ?",
            (appointment_id,)
        )
        return {"ok": True}
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))
